### DISCUSSION TITLE
How would I use the domains dynamic factory if I have multiple domains that are tied to the same SObject?

### DISCUSSION THREAD

**Discussion by daniel-fuller** [wimvelzeboer/fflib-apex-extensions#43] [Q&A] [answered]
@wimvelzeboer, I was looking at `fflib_DynamicDomainFactory` and was wondering what changes I would need to make if I have multiple domains associated with a single SObject? For example, in our codebase we have different Account types that each have a large set of unique behaviors so we have separated those into separate domains. However, I see in the dynamic factory `System.Type domainImplementationType = bindings.get(0).getImplementationType();` so if there are multiple implementations tied to the object type grouping and only the 0 index binding is retrieved, it appeared to me that having multiple implementations/interfaces tied to an object type wouldn't work. Is my thinking correct there? If so, is there a way I could extend the current functionality to allow for that?

**Comment by wimvelzeboer** [comment]
@dfuller114 I think you would love if Salesforce would make their SObject inheritance database available, (where you can extend an existing table, making recordtypes virtually obsolete), but we are not there yet.

You are indeed right that by default one SObjectType has one domain implementation. In my projects I typically tried to split the application into pieces e.g.; Community App / Sales App  / Service App. Each App has therefore its own Application class and its own domain class.

But if you then still have many methods on the domain and desire to split the domain per record type. I would create one domain, with static methods doing something like:
```
public static IPartnerAccounts newPatnerAccounts(List<SObject> records)
{
    Application.Domain.replaceWith(Accounts.SObjectType, PartnerAccounts.Constructor.class);
    (IPartnerAccounts) application.Domain.newInstance(records);
}
```
Or you can utilise a Service Factory to configure the domain constructors, 
```
public class Accounts extends fflib_SObjects2 
{
    ...
    public static IPartnerAccounts newPartnerAccounts(List<SObject> records)
    {
        PartnerAccounts.Constructor constructor= 
                (PartnerAccounts.Constructor) Application.ComplexDomain.newInstance(IPartnerAccounts.class);

        return constructor.construct(records);
    }
    ...
}
```
```
public class Application
{
     public static final fflib_IServiceFactory ComplexDomain =
            new fflib_ClassicServiceFactory(
                    new Map<Type, Type>
                    {
                            IPartnerAccounts.class => PartnerAccounts.Constructor.class 
                    }
            );
}
```
you can then use `IPartnerAccounts domain = Accounts.newPartnerAccounts(records);` to construct it.

But I would avoid it in general, as something like this would fail if (some) of the provided records where not of that RecordType.

**Comment by daniel-fuller** [comment]
@wimvelzeboer Thanks for that thorough info. Really helpful. This might just be a knowledge gap on my part as I've just started using the library recently. I noticed the `Constructor` class reference in `PartnerAccounts.Constructor.class`. Wasn't sure what that reference's purpose was. 

Also, I don't see `replaceWith()` implemented on the DynamicDomainFactory like with the DynamicServiceFactory. Is that something I would need to implement myself? I assume it would be similar to the `replaceWith()` in the in the DynamicServiceFactory. Because in this case, I would have different domain interfaces mapped to different implementation types. For example, we have IAccounts => Accounts, IHouseholdAccounts => IHouseholdAccounts, etc. At least that's how I am envisioning it. 

So I was thinking of having a similar block like the one below, but I'm not entirely sure where the domain SObject record collection comes into play here.

```
public void replaceWith(SObjectType sObjectType, Type domainInterfaceType, Type replacementImplType) {
		resolver.byType(fflib_AppBindingType.Domain)
				.byObject(sObjectType)
				.byInterfaceType(domainInterfaceType)
				.resolve()
				.get(0)
				.setImplementationType(replacementImplType);
	}
```

**Comment by wimvelzeboer** [comment, marked as answer]
@dfuller114 The Constructor sub-class is something all domains should have to overcame an Apex limitation. Apex is currently not able to construct a class dynamically when the constructor method requires a parameter. This Constructor sub-class will overcome that limitation.

But I had another question for you about the size of the domains. Are your domains extended from fflib_SObjects/fflib_SObjects2 or fflib_SObjectDomain? With other words, do you still mix trigger functionality in your domain or did you already which to a pure Domain?
If you still have trigger functionality in the domain, then you would want to split that so that you can limit the size of the domain file. 
Try extendind the domain from fflib_SObjects2 and the triggerhandler from fflib_SObjectTriggerHandler.

> Also, I don't see replaceWith() implemented on the DynamicDomainFactory like with the DynamicServiceFactory. Is that something I would need to implement myself? I assume it would be similar to the replaceWith() in the in the DynamicServiceFactory. Because in this case, I would have different domain interfaces mapped to different implementation types. For example, we have IAccounts => Accounts, IHouseholdAccounts => IHouseholdAccounts, etc. At least that's how I am envisioning it.

It seems that the `replaceWith()` method was removed when it was migraded to fflib-apex-commons, thats a bit unfortunate. Should add that one again..
